<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>shaman - Sampling HiC contAct Matrices for Aparametric Normalization • shaman</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="shaman - Sampling HiC contAct Matrices for Aparametric Normalization">
<meta property="og:description" content="Generate expected HiC data maintaining marginal coverage,
  and the decay in contact probability with increasing genomic distance. Expected
  HiC data is used for normalization and graphical presenstation of statistical
  significance of observed HiC contact enrichments.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">shaman</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="articles/shaman-package.html">
    <span class="fa fa-road"></span>
     
    Usage
  </a>
</li>
<li>
  <a href="articles/import.html">
    <span class="fa fa-filter"></span>
     
    Import
  </a>
</li>
<li>
  <a href="reference/index.html">
    <span class="fa fa-th-list"></span>
     
    Reference
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.org/tanaylab/shaman" class="external-link">
    <span class="fa fa-github-square"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="the-shaman-package---sampling-hic-contact-matrices-for-a-parametric-normalization">The shaman package - sampling HiC contact matrices for a-parametric normalization<a class="anchor" aria-label="anchor" href="#the-shaman-package---sampling-hic-contact-matrices-for-a-parametric-normalization"></a>
</h1></div>
<p>Shaman package impelements functions for resampling Hi-C matrices in order to generate expected contact distributions given constraints on marginal coverage and contact-distance probability distributions. The package also provides support for visualizing normalized matrices and statistical analysis of contact distributions around selected landmarks. It is an R package embedding algorithms implemened in C++, which is built over support from the Tanay’s lab <em>Misha</em> genomic database system that is provided with the package.</p>
<p>The normalization workflow consists of the following steps:</p>
<ol>
<li>
<a href="https://tanaylab.github.io/shaman/articles/import.html" class="external-link">Import</a> observed HiC data to misha db. In order to utilize full functionality of the shaman package, you need to construct a Misha database for your genome of interest, and import your Hi-C data to Misha binary Quad-tree format.</li>
<li>Shuffle observed data to generate expected data. Reshuffling of an entire dataset will require 7 hours per 1 billion reads on a machine with one core per chromosome. Note that shaman can also run distributed using Sun Grid Engine.</li>
<li>Compute the normalized score of each observed data point. This computation can be conducted inline for a specific genomic region, or pre-processed on the entire dataset using distributed computation. Score computation on 1 billion reads on a distributed system may take 4-10 hours, depending on the number of cores available.</li>
<li>Visualize normalized contact maps. Precomputed normalized scores allow for flexible analysis of contact distributions.</li>
<li>Quantify contact enrichment acoording to genomic features.</li>
</ol>
<p>In the example misha database provided in this package we have created a low-footprint matrix to examplify the shaman workflow. We included 4.6 million contacts from ELA K562 dataset covering the hoxd locus (chr2:175e06-178e06) and convergent CTCF regions. Processing the complete matrix from this study requires downloading the full contact list and regenerating the reshuffled matrix.</p>
<div class="section level3">
<h3 id="code">Code<a class="anchor" aria-label="anchor" href="#code"></a>
</h3>
<p>Source code can be found at: <a href="https://github.com/tanaylab/shaman" class="external-link uri">https://github.com/tanaylab/shaman</a></p>
</div>
<div class="section level3">
<h3 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h3>
<ul>
<li><em>Perl</em></li>
<li>
<em>devtools</em> R package (optional, for automatic installation of <em>bioconductor</em> dependencies)</li>
<li>
<em>System:</em> multi-core unix / linux based system or SGE (sun grid engine) cluster for distributed computing are required for large HiC datasets.</li>
</ul>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>The quickest way to install <em>shaman</em> is to use the following command:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'tanaylab/shaman'</span>, 
                            vignette <span class="op">=</span> <span class="cn">TRUE</span>, 
                            repos<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">'repos'</span><span class="op">)</span>, <span class="st">'https://tanaylab.github.io/repo'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>If devtools fails to install all the bioconductor requirments please install <em>Gviz</em> and <em>GenomeInfoDb</em> manually from bioconductor:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"https://bioconductor.org/biocLite.R"</span><span class="op">)</span>
<span class="fu">biocLite</span><span class="op">(</span><span class="st">"Gviz"</span><span class="op">)</span>
<span class="fu">biocLite</span><span class="op">(</span><span class="st">"GenomeInfoDb"</span><span class="op">)</span></code></pre></div>
<div class="section level4">
<h4 id="installing-without-the-devtools-package">Installing without the <em>devtools</em> package<a class="anchor" aria-label="anchor" href="#installing-without-the-devtools-package"></a>
</h4>
<p>In order to install without the <em>devtools</em> package, please take the following steps:</p>
<ol>
<li>Install <em>Gviz</em> and <em>GenomeInfoDb</em> from <em>bioconductor</em>:</li>
</ol>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"https://bioconductor.org/biocLite.R"</span><span class="op">)</span>
<span class="fu">biocLite</span><span class="op">(</span><span class="st">"Gviz"</span><span class="op">)</span>
<span class="fu">biocLite</span><span class="op">(</span><span class="st">"GenomeInfoDb"</span><span class="op">)</span></code></pre></div>
<ol>
<li>Install <em>shaman</em>:</li>
</ol>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'shaman'</span>, repos<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">'repos'</span><span class="op">)</span>, <span class="st">'https://tanaylab.github.io/repo'</span><span class="op">)</span><span class="op">)</span> </code></pre></div>
</div>
<div class="section level4">
<h4 id="when-all-else-fails">When all else fails:<a class="anchor" aria-label="anchor" href="#when-all-else-fails"></a>
</h4>
<p>In order to install from source, please take the following steps:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"http://www.wisdom.weizmann.ac.il/~nettam/shaman/misha_3.5.6.tar.gz"</span>, , repos<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span> <span class="co"># Download and install misha package) </span>
<span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"https://bioconductor.org/biocLite.R"</span><span class="op">)</span> <span class="co">#installing Gviz</span>
<span class="fu">biocLite</span><span class="op">(</span><span class="st">"Gviz"</span><span class="op">)</span>
<span class="fu">biocLite</span><span class="op">(</span><span class="st">"GenomeInfoDb"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"http://www.wisdom.weizmann.ac.il/~nettam/shaman/shaman_2.0.tar.gz"</span>, , repos<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span> <span class="co"># Download and install shaman package)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="using-the-package">Using the package<a class="anchor" aria-label="anchor" href="#using-the-package"></a>
</h3>
<p>Please refer to <a href="https://tanaylab.github.io/shaman/articles/shaman-package.html" class="external-link uri">https://tanaylab.github.io/shaman/articles/shaman-package.html</a> for usage and workflow.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/tanaylab/shaman/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/tanaylab/shaman/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>GPL</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing shaman</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Amos Tanay <br><small class="roles"> Author </small>  </li>
<li>Netta Mendelson Cohen <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Amos Tanay, Netta Mendelson Cohen.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
