<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>plots hic normalized map with annotations — shaman_plot_map_score_with_annotations • shaman</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="plots hic normalized map with annotations — shaman_plot_map_score_with_annotations"><meta property="og:description" content="shaman_plot_map_score_with_annotations"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">shaman</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/shaman-package.html">
    <span class="fa fa-road"></span>
     
    Usage
  </a>
</li>
<li>
  <a href="../articles/import.html">
    <span class="fa fa-filter"></span>
     
    Import
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-th-list"></span>
     
    Reference
  </a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.org/tanaylab/shaman" class="external-link">
    <span class="fa fa-github-square"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>plots hic normalized map with annotations</h1>
    <small class="dont-index">Source: <a href="https://github.com/tanaylab/shaman/blob/HEAD/R/shaman_plot_map.R" class="external-link"><code>R/shaman_plot_map.R</code></a></small>
    <div class="hidden name"><code>shaman_plot_map_score_with_annotations.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>shaman_plot_map_score_with_annotations</code></p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">shaman_plot_map_score_with_annotations</span><span class="op">(</span>
  <span class="va">genome</span>,
  <span class="va">points_score</span>,
  <span class="va">interval_range</span>,
  point_size <span class="op">=</span> <span class="fl">0.1</span>,
  misha_tracks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,
  mt_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"shaman.track_colors"</span><span class="op">)</span>,
  mt_ylims <span class="op">=</span> <span class="cn">NULL</span>,
  annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,
  a_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"shaman.annotation_colors"</span><span class="op">)</span>,
  add_genes <span class="op">=</span> <span class="cn">T</span>,
  add_ideogram <span class="op">=</span> <span class="cn">T</span>,
  add_axis <span class="op">=</span> <span class="cn">T</span>,
  gene_stacking <span class="op">=</span> <span class="st">"squish"</span>,
  gene_size <span class="op">=</span> <span class="fl">0.7</span>,
  track_size <span class="op">=</span> <span class="fl">0.8</span>,
  annotation_size <span class="op">=</span> <span class="fl">0.7</span>,
  fig_fn <span class="op">=</span> <span class="st">""</span>,
  fig_width <span class="op">=</span> <span class="fl">900</span>,
  fig_height <span class="op">=</span> <span class="fl">5</span><span class="op">/</span><span class="fl">6</span> <span class="op">*</span> <span class="fl">900</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>genome</dt>
<dd><p>Name of reference genome (e.g. "hg19", "mm10")</p></dd>
<dt>points_score</dt>
<dd><p>A dataframe containing the points (start1, start2) and their normalized score.</p></dd>
<dt>interval_range</dt>
<dd><p>1D interval (chrom, start, end) specifying the region to plot</p></dd>
<dt>misha_tracks</dt>
<dd><p>List of 1D track expressions (virtual tracks also supported) which can be extracted from.</p></dd>
<dt>mt_colors</dt>
<dd><p>Array of colors, one for each misha_track, which will be used to plot each 1d track.</p></dd>
<dt>annotations</dt>
<dd><p>List of gintervals highlighting annotated regions.</p></dd>
<dt>a_colors</dt>
<dd><p>Array of colors, one for each annotation set.</p></dd>
<dt>add_genes</dt>
<dd><p>Boolean flag indicating whether to show gene annotations.</p></dd>
<dt>add_ideogram</dt>
<dd><p>Boolean flag indicating whether to add an chromosomal ideogram.</p></dd>
<dt>add_axis</dt>
<dd><p>Boolean flag indicating whether to show chromosmal axis.</p></dd>
<dt>gene_stacking</dt>
<dd><p>Describes the viewing option of the genes track. Can be either "squish", or "dense".</p></dd>
<dt>gene_size</dt>
<dd><p>Size of gene annotation view.</p></dd>
<dt>track_size</dt>
<dd><p>Size of track view.</p></dd>
<dt>fig_fn</dt>
<dd><p>Name of png file to output to. Empty string will cause the figure to be plotted to the current device.</p></dd>
<dt>fig_width</dt>
<dd><p>Width in pixels of output png figure.</p></dd>
<dt>fig_height</dt>
<dd><p>Height in pixels of output png figure.</p></dd>
<dt>my_ylims</dt>
<dd><p>Y-axis limits for each misha track. If not provided, generated automatically based on the data
to be displayed in the region.</p></dd>
<dt>annot_size</dt>
<dd><p>Size of annotation view.</p></dd>
</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>Plots observerved hic contact matrix color-coded by normalized scores, aligned with
linear data such as gene annotations, rna, chromatin, transcription factor binding, etc').</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"></span>
<span class="r-in"><span class="fu">gsetroot</span><span class="op">(</span><span class="fu"><a href="shaman_get_test_track_db.html">shaman_get_test_track_db</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in setwd(groot):</span> cannot change working directory</span>
<span class="r-in"><span class="va">points</span> <span class="op">&lt;-</span> <span class="fu">gextract</span><span class="op">(</span><span class="st">"hic_score"</span>, <span class="fu">gintervals.2d</span><span class="op">(</span><span class="fl">2</span>, <span class="fl">175e06</span>, <span class="fl">178e06</span>, <span class="fl">2</span>, <span class="fl">175e06</span>, <span class="fl">178e06</span><span class="op">)</span>, colnames <span class="op">=</span> <span class="st">"score"</span><span class="op">)</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> Invalid interval (chr2, 1.75e+08, 1.78e+08): end coordinate exceeds chromosome boundaries</span>
<span class="r-in"><span class="fu">shaman_plot_map_score_with_annotations</span><span class="op">(</span><span class="st">"hg19"</span>, <span class="va">points</span>, <span class="fu">gintervals</span><span class="op">(</span><span class="fl">2</span>, <span class="fl">175e06</span>, <span class="fl">178e06</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in shaman_plot_map_score_with_annotations("hg19", points, gintervals(2,     175000000, 178000000)):</span> points_score data frame must contain the following columns: start1, start2, score</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Amos Tanay, Netta Mendelson Cohen.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

